apps/condo/domains/acquiring/tasks/recurrent-payments/recurrent-payment-processing.js

```diff
@@ -0,0 +1,128 @@
+const { get } = require('lodash')
+
+const { getLogger } = require('@open-condo/keystone/logging')
+const { getSchemaCtx } = require('@open-condo/keystone/schema')
+
+const {
+    paginationConfiguration,
+    PAYMENT_ERROR_UNKNOWN_CODE,
+    PAYMENT_ERROR_CARD_TOKEN_NOT_VALID_CODE,
+} = require('@condo/domains/acquiring/tasks/utils/constants')
+const {
+    PaymentAdapter,
+} = require('@condo/domains/acquiring/tasks/utils/PaymentAdapter')
+const {
+    getReadyForProcessingPaymentsPage,
+    registerMultiPayment,
+    setRecurrentPaymentAsFailed,
+    setRecurrentPaymentAsSuccess,
+} = require('@condo/domains/acquiring/tasks/utils/queries')
+const {
+    RecurrentPaymentContext,
+} = require('@condo/domains/acquiring/utils/serverSchema')
+const { processArrayOf } = require('@condo/domains/common/utils/parallel')
+
+const logger = getLogger('recurrent-payment-context-processing')
+
+async function processRecurrentPayment (context, recurrentPayment, paymentAdapter) {
+    // prepare vars
+    const { recurrentPaymentContext: { id: recurrentContextId } } = recurrentPayment
+    const { settings: { cardId } } = await RecurrentPaymentContext.getOne(context, {
+        id: recurrentContextId,
+    })
+
+    // check card token validity
+    const cardTokenValidity = await paymentAdapter.checkCardToken(cardId)
+    if (!cardTokenValidity) {
+        return setRecurrentPaymentAsFailed(
+            context,
+            recurrentPayment,
+            `Provided card token id is not valid ${cardId}`,
+            PAYMENT_ERROR_CARD_TOKEN_NOT_VALID_CODE,
+        )
+    }
+
+    // proceed payment
+    const { paid, errorMessage, errorCode } = await paymentAdapter.proceedPayment(cardId)
+
+    // update status
+    if (paid) {
+        return await setRecurrentPaymentAsSuccess(context, recurrentPayment)
+    } else {
+        return await setRecurrentPaymentAsFailed(context, recurrentPayment, errorMessage, errorCode)
+    }
+}
+
+async function process () {
+    logger.info({ msg: 'Start processing recurrent payment tasks' })
+
+    // prepare context
+    const { keystone } = await getSchemaCtx('RecurrentPaymentContext')
+    const context = await keystone.createContext({ skipAccessControl: true })
+
+    // prepare vars
+    const { pageSize } = paginationConfiguration
+    let offset = 0
+    let hasMorePages = true
+
+    // retrieve RecurrentPaymentContext page by page
+    while (hasMorePages) {
+        logger.info({ msg: `Processing recurrent payment page #${Math.floor(offset / pageSize)}` })
+        // get page (can be empty)
+        const page = await getReadyForProcessingPaymentsPage(context, pageSize, offset)
```

Comment:
We don't have some paging util? It looks like topical task

Reply:
We don't have such util (scan an entity, page by page and proceed page items in parallel)